src/Adapter/Order/CommandHandler/AddProductToOrderHandler.php

```diff
@@ -89,132 +99,137 @@ public function handle(AddProductToOrderCommand $command)
+
+            StockAvailable::synchronize($product->id);
 
-        $order = $order->refreshShippingCost();
+            // Update weight SUM
+            $orderCarrier = new OrderCarrier((int) $order->getIdOrderCarrier());
+            if (Validate::isLoadedObject($orderCarrier)) {
+                $orderCarrier->weight = (float) $order->getTotalWeight();
+                if ($orderCarrier->update()) {
+                    $order->weight = sprintf('%.3f ' . Configuration::get('PS_WEIGHT_UNIT'), $orderCarrier->weight);
+                }
+            }
 
-        Hook::exec('actionOrderEdited', ['order' => $order]);
+            // Update Tax lines
+            $orderDetail->updateTaxAmount($order);
 
-        $newCartRules = $this->context->cart->getCartRules();
+            // Delete specific price if exists
+            if (null !== $specificPrice) {
+                $specificPrice->delete();
+            }
 
-        sort($oldCartRules);
-        sort($newCartRules);
+            $order = $order->refreshShippingCost();
+
+            Hook::exec('actionOrderEdited', ['order' => $order]);
+
+            $newCartRules = $this->context->cart->getCartRules();
+
+            sort($oldCartRules);
+            sort($newCartRules);
+
+            $result = !empty($newCartRules) && !empty($oldCartRules) ? array_diff($newCartRules, $oldCartRules) : [];
+
+            foreach ($result as $cartRule) {
+                // Create OrderCartRule
+                $rule = new CartRule($cartRule['id_cart_rule']);
+                $values = [
+                    'tax_incl' => $rule->getContextualValue(true),
+                    'tax_excl' => $rule->getContextualValue(false),
+                ];
+                $orderCartRule = new OrderCartRule();
+                $orderCartRule->id_order = $order->id;
+                $orderCartRule->id_cart_rule = $cartRule['id_cart_rule'];
+                $orderCartRule->id_order_invoice = $invoice->id;
+                $orderCartRule->name = $cartRule['name'];
+                $orderCartRule->value = $values['tax_incl'];
+                $orderCartRule->value_tax_excl = $values['tax_excl'];
+                $orderCartRule->add();
+
+                // @todo: use https://github.com/PrestaShop/decimal
+                $order->total_discounts += $orderCartRule->value;
+                $order->total_discounts_tax_incl += $orderCartRule->value;
+                $order->total_discounts_tax_excl += $orderCartRule->value_tax_excl;
+                $order->total_paid -= $orderCartRule->value;
+                $order->total_paid_tax_incl -= $orderCartRule->value;
+                $order->total_paid_tax_excl -= $orderCartRule->value_tax_excl;
+            }
 
-        if (!empty($newCartRules) && !empty($oldCartRules)) {
-            $result = array_diff($newCartRules, $oldCartRules);
-        } else {
-            $result = [];
-        }
+            $order->update();
+        } catch (Exception $e) {
```

Comment:
The exception is too generic :thinking:

Reply:
Not in that case, the idea is just to:
- catch the exception
- restore the context
- rethrow the exception

So it is voluntary that it's the most generic one